proposals/NNNN-max-records-per-node.md

```diff
@@ -0,0 +1,114 @@
+
+**Developer's note**: Implementations that do not support or ignore this attribute, will not be functionally impacted.
+
+### Usage
+
+The following trivial example demonstrates using `MaxRecordsPerNode` for a thread launch node which distributes
+a single record across an array of 64 consumer thread launch nodes.
+
+```cpp
+[Shader("node")]
+[NodeLaunch("thread")]
+[NodeIsProgramEntry]
+void DispatchNode(
+    [MaxRecords(64)]       // a maximum of 64 records are written to output node array,
+    [MaxRecordsPerNode(1)] // but only 1 record is written to each node in the array
+    [NodeArraySize(64)] NodeOutputArray<RECORD> ConsumerNodes )
+{
+    [unroll] for(uint i = 0; i < 64; ++i)
+    {
+        ThreadNodeOutputRecords<RECORD> outputRecord = ConsumerNodes[i].GetThreadNodeOutputRecords(1);
+        ...
+        outputRecord.OutputComplete();
+    }
+}
+```
+
+As mentioned above, some material shading algorithms have a similar pattern: a single node which makes a decision about
+which node(s) in a node array (materials) to execute, where the number of possible materials is large, but the number of
+records submitted to any specific node is small, relative to the size of the array.
+
+### Interchange Format Additions
+
+A new metadata tag is added for MaxRecordsPerNode.
+
+|Tag                            |Tag Encoding     |Value Type     |Default     |
+|:------------------            |:----------------|:--------------|:-----------|
+|kDxilNodeMaxRecordsPerNodeTag  |`7`              |`i32`          |Required, See [HLSL Additions](#hlsl-additions) section for backward compatibility with older Shader Models    |  
+
+### Runtime Additions
+
+The `MaxRecordsPerNode` information will be captured to RDAT. Similar to other Node attributes, add a `RDAT::NodeAttribKind` named `MaxRecordsPerNode`.
+
+## Alternatives considered
+
+### Parameter of MaxRecords
+
+Modify the definition for `MaxRecords` node output attribute:
+
+| attribute | required | description |
+|:---       |:--------:|:------------|
+| `[MaxRecords(count)]` or `[MaxRecords(count, maxRecordsPerNode)]` | Y (this or below attribute) | Given uint `count` declaration, the thread group can output `0...count` records to this output.  The variant with `maxRecordsPerNode` is required for `NodeArrayOutput`, where `count` applies across all the output nodes in the array and `maxRecordsPerNode` specifies the maximum number of records that can be written to a single output node within the array.  Exceeding these limits results in undefined behavior.  The value of `maxRecordsPerNode` must be less-than or equal to the value of `count`.  These attributes can be overridden via the `NumOutputOverrides / pOutputOverrides` option when constructing a work graph as part of the [definition of a node]().  See [Node output limits](). |
```

Comment:
`MaxRecordsPerNode` is defined as a separate attribute earlier on, instead of an extra parameter to the existing `MaxRecords` attribute.  Which is it?

Reply:
The proposal is a separate attribute.  This section shows an "alternative" method as an extra parameter, which was earlier discussed in other forums.

Reply:
@tex3d Better?

Reply:
Apologies for the confusion from before.  For some reason when I was reading these I wasn't realizing they were under an "alternative" section.